### DISCUSSION TITLE
Sync Across Multiple Devices

### DISCUSSION THREAD

**Discussion by rosie-at-pieces** [pieces-app/support#29] [General]
Hi everyone! We've been hearing a lot of requests for the ability to sync your snippets and copilot chats across multiple devices. We agree that this is a really important feature and are investigating the best way to do it. In the meantime, we have recently [published a guide](https://docs.pieces.app/product-highlights-and-benefits/sync-across-multiple-devices-using-github) on how to do this manually. It's a beta process so please don't hesitate to [contact us](https://getpieces.typeform.com/to/mCjBSIjF#page=docs-sidebar) if you are having issues and we'll help you out!

In the meantime, we'd love to hear from people on if they've tried it out and how successful they were, any tips/tricks they learned while doing it, etc. Post your input below!

**Comment by peaklabs-dev** [comment]
I would love a cloud sync across multiple devices. Also make sure the cloud storage is encrypted and we can use your own cloud storage like SFTP, Backblaze or OneDrive.

**Comment by yahiaaktham** [comment]
Following up on this thread two and a half years later — cross-device Long-Term Memory is still the single biggest gap for multi-machine users, and I want to describe the concrete shape of the need rather than just .

My setup: two Windows laptops — one corporate (heavily restricted, Cursor + MCP), one personal. Both run PiecesOS, both signed into the same account, both with Cloud connected.

What I expected: signing into the same account on both machines would give my agents one continuous LTM.

What actually happens: each install builds a completely independent LTM. ask_pieces_ltm on laptop A has no knowledge of anything that happened on laptop B. The cloud connection appears to cover account-level items (preferences, saved materials), not captured workstream memory.

Why the current workarounds don't solve it: the two available options — copying the com.pieces.os database folder, and Personal Cloud Backup & Restore — are both replace operations. Restoring laptop A's snapshot onto laptop B destroys B's own memory. For anyone who genuinely works on both machines, that's unusable: I'd be permanently choosing which half of my work history to delete.

What I'm asking for — an additive merge, not a mirror:

Opt-in cross-device LTM sync where workstream summaries from all of a user's devices coexist in one searchable timeline.

Device provenance on each memory (a source_device field), plus a filter in Workstream Activity so I can scope recall to one machine when I want to.

Union-on-conflict semantics — never delete or overwrite a memory that exists only on the local device.

Per-device opt-out, so a restricted corporate machine can be excluded for compliance reasons.

Failing all of the above: a documented, supported import path (append summaries + annotations without wiping existing ones) so the community can build this reliably against a stable contract.

On the API workaround: I'm currently prototyping a bidirectional merge script that reads workstream summaries from the local PiecesOS REST API on each laptop, ships JSONL through a shared bucket, and re-creates only the missing ones with a [from: <device>] prefix and a hash-based dedup ledger. It's additive and it works in principle, but it's fragile — the endpoint shapes aren't guaranteed across PiecesOS versions, and injected summaries aren't equivalent to natively captured context. I'd much rather this were a first-class feature.

Questions for the team:

Is additive multi-device LTM on the roadmap, and is there a target release?

Are the workstream-summary create/annotation endpoints considered stable enough to build a community sync tool on, or should I expect breaking changes?

Is there any currently supported way to append memory from an export rather than restore-and-replace?

Happy to test any early-access build and report back with logs. PiecesOS version and OS details available on request.

**Reply by mark-at-pieces** [reply]
Hey @yahiaaktham you will be happy to know this is actively in development.

There are a few pillars that we are cleaning up that are precursors, but TLDR: we are actively working on 2 cloud-related features, peer-to-peer sharing + realtime sync(for all memories). Expected to land End of August/mid-September!

That being said api design aligns with PiecesOS versions. Our endpoints are very stable at this point; it has been over a year since our last breaking change, and we do not have any breaking changes planned anytime soon!

There is 100% a route to import/export memories today, but it is a bit rough and dirty, and would require pairing a good deal of endpoints.

Hopefully, this helps provide some clarity!
